src/parser/shadowlands/modules/items/dungeons/ShadowgraspTotem.tsx

```diff
@@ -0,0 +1,205 @@
+ * - https://www.warcraftlogs.com/reports/twQhDxWCZN2jPbnp#fight=6&type=damage-done&source=13
+ */
+class ShadowgraspTotem extends Analyzer {
+  static dependencies = {
+    abilities: Abilities,
+    castEfficiency: CastEfficiency,
+    spellUsable: SpellUsable,
+  };
+  protected abilities!: Abilities;
+  protected castEfficiency!: CastEfficiency;
+  protected spellUsable!: SpellUsable;
+
+  private item: Item;
+
+  /** Timestamps of damage events since the last cast */
+  private currentDot?: number[];
+  /** All timestamps of all casts */
+  private dots: number[][] = [];
+  private totalDamage = 0;
+  private healing: number[] = [];
+
+  constructor(
+    options: Options & {
+      abilities: Abilities;
+      castEfficiency: CastEfficiency;
+      spellUsable: SpellUsable;
+    },
+  ) {
+    super(options);
+
+    this.item = this.selectedCombatant.getItem(ITEM.id)!;
+    if (this.item == null) {
+      this.active = false;
+      return;
+    }
+
+    // Add cast as an ability to show effective usage and cooldown in timeline
+    options.abilities.add({
+      spell: SUMMON_SPELL.id,
+      category: Abilities.SPELL_CATEGORIES.ITEMS,
+      cooldown: DEFAULT_COOLDOWN_SECONDS,
+      gcd: null,
+      castEfficiency: {
+        suggestion: true,
+        recommendedEfficiency: 0.8,
+      },
+    });
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SUMMON_SPELL), this.onCast);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER_PET).spell(DAMAGE_SPELL), this.onDamage);
+    this.addEventListener(Events.heal.to(SELECTED_PLAYER).spell(HEAL_SPELL), this.onHeal);
+  }
+
+  onCast(event: CastEvent) {
+    this.dots.push((this.currentDot = []));
+  }
+
+  onDamage(event: DamageEvent) {
+    this.currentDot?.push(event.timestamp);
+    this.totalDamage += (event.amount || 0) + (event.absorbed || 0);
+  }
+
+  onHeal(event: HealEvent) {
+    this.healing.push((event.amount || 0) + (event.absorbed || 0));
+
+    // Inform other modules that cooldown has been reduced
+    if (this.spellUsable.isOnCooldown(SUMMON_SPELL.id)) {
+      this.spellUsable.reduceCooldown(SUMMON_SPELL.id, COOLDOWN_REDUCTION_SECONDS * 1000);
+    }
+  }
+
+  private getCastEfficiancy() {
+    return this.castEfficiency.getCastEfficiencyForSpellId(SUMMON_SPELL.id);
+  }
+
+  private get numberHeals() {
+    return this.healing.length;
+  }
+
+  private get totalHealing() {
+    return this.healing.reduce((total, healing) => total + healing, 0);
+  }
+
+  private get averageHealing() {
+    return this.totalHealing / this.numberHeals;
+  }
+
+  private get averageTickRate() {
+    return this.dots
+      .map((timestamps) =>
+        timestamps.reduce(
+          (deltas, timestamp, i, array) =>
+            i === 0 ? deltas : deltas.concat(timestamp - array[i - 1]),
+          [] as number[],
+        ),
+      )
+      .flat()
+      .reduce((average, tickRate, i, array) => average + tickRate / array.length, 0);
+  }
```

Comment:
not convinced that knowing the avg tick rate is useful, but I don't see any harm in having it

Reply:
I use it to calculate an "effective haste". The point is that if you use it during haste cds, like during invoker's delight, it'll do more damage. So just thought it'd be nice way to provide some feedback for maximum benefit of it.

Reply:
I use it to calculate an "effective haste". The point is that if you use it during haste cds, like during invoker's delight, it'll do more damage. So just thought it'd be nice way to provide some feedback for maximum benefit of it.